File: frontend/aura/api/public/authors.ts
import { type Ref, unref } from "vue";
import type { AuthorArchive, PostListResponse } from "../contracts/shared";
import {
	type ApiQueryOptions,
	type ApiQueryPath,
	type QueryParams,
	query,
	withQuery,
} from "../transport";

type Getter<T> = () => T;
type MaybeGetter<T> = T | Getter<T> | Ref<T>;

/** A public author's archive (DEC-359/TASK-405): the author envelope rides on
 *  the (possibly empty) post list so the page can title itself by pen name
 *  even before the writer has published anything. The envelope carries the
 *  archive-shaped author — AuthorArchive (round 357) — with the "about this
 *  writer" bio; per-post lists keep the slim AuthorBrief. */
export interface AuthorPostsResponse extends PostListResponse {
	author?: AuthorArchive | null;
}

/** A writers-index row (rounds 346/357): pen name + published-post count +
 *  the public "about this writer" bio (null until a superuser writes one). */
export interface AuthorIndex {
	id: number;
	display_name: string;
	post_count: number;
	bio?: string | null;
}

/** Every public writer, pen name + published-post count (round 346) — powers
 *  the /authors index page. Anonymous + caching-friendly. */
export function useAuthors() {
	return query<AuthorIndex[]>("/api/authors");
}

export interface AuthorPostFilters {
	page?: number;
	limit?: number;
}

/** A writer's published posts, newest-first (paginated).
 *
 * The backend answers 404 for unknown ids and for admins with no pen name (no
 * public presence — the username never surfaces), so a 404 here IS the page's
 * "author not found" signal. Pass a getter so useFetch refetches on SPA
 * navigation between authors.
 */
export function useAuthorPosts(
	authorId: string | number | Getter<string | number | null | undefined>,
	filters: MaybeGetter<AuthorPostFilters> = {},
	options: ApiQueryOptions<AuthorPostsResponse> = {},
) {
	const path = (() => {
		const id = typeof authorId === "function" ? authorId() : unref(authorId);
		if (id == null) return null;
		// AuthorPostFilters is structurally a QueryParams superset; the cast
		// keeps the caller-facing type precise while withQuery works on the
		// generic record (same cast as usePosts' filters).
		const params = (typeof filters === "function" ? filters() : unref(filters)) as QueryParams;
		return withQuery(`/api/authors/${id}/posts`, params);
	}) as ApiQueryPath;
	return query<AuthorPostsResponse>(path, options);
}

File: backend/nova/app/routers/authors.py
"""Public author archive (DEC-359, TASK-405).

An admin writer with a public pen name (User.display_name) has a public
presence: their byline on posts, and this per-author archive of published
posts. The login username is never exposed — admin login is deliberately
no-oracle (timing-equalized; see admin login) — and an unknown author id
answers 404 exactly like an author who chose no pen name, so this endpoint can
neither enumerate admins nor reveal which usernames exist.
"""

from fastapi import APIRouter, Depends, HTTPException, Query, Request
from sqlalchemy import func
from sqlalchemy.orm import Session

from app import auth, crud, models, schemas
from app.conditional import conditional_json
from app.database import get_db
from app.schemas import IdInt, PageInt

router = APIRouter(prefix="/api/authors", tags=["authors"])


@router.get("", response_model=list[schemas.AuthorIndex])
def list_authors(db: Session = Depends(get_db)):
    """Every public writer (pen-named admin), with their published-post count.

    Writer discovery (DEC-359, round 346): a reader who found one byline can
    browse every contributor. Only pen names are exposed — an admin with no
    pen name has no public presence and stays off this list (never the login
    username, admin login is no-oracle). The count is published posts only, so
    a writer with nothing live yet still appears with a zero.
    """
    writers = db.query(auth.User).filter(auth.User.display_name.isnot(None)).order_by(auth.User.display_name).all()
    post_counts = (
        db.query(models.Post.author_id, func.count(models.Post.id))
        .filter(models.Post.author_id.isnot(None), models.Post.published.is_(True))
        .group_by(models.Post.author_id)
        .all()
    )
    counts = {author_id: int(count) for author_id, count in post_counts}
    items = [
        {
            "id": u.id,
            "display_name": u.display_name,
            "post_count": counts.get(u.id, 0),
            "bio": u.bio,
        }
        for u in writers
    ]
    return items


@router.get("/{author_id}/posts", response_model=schemas.PostListResponse)
def get_author_posts(
    request: Request,
    author_id: IdInt,
    page: PageInt = 1,
    limit: int = Query(10, ge=1, le=50),
    db: Session = Depends(get_db),
):
    """A writer's published posts, newest-first (paginated).

    Only reachable for a public author — an admin who set a pen name. An
    unknown id and a username-only admin answer the same 404 (no public
    presence), so the endpoint neither enumerates admins nor leaks login
    usernames (DEC-359, TASK-405).
    """
    # Fire the scheduled-post publish-time fan-out on this surface too (the
    # round-331 sweep: every surface that can show a crossed scheduled post
    # announces it exactly once).
    crud.maybe_notify_due_scheduled_posts(db)
    author = db.query(auth.User).filter(auth.User.id == author_id).first()
    if author is None or not author.display_name:
        raise HTTPException(status_code=404, detail="Author not found")

    posts, total = crud.get_posts(db, skip=(page - 1) * limit, limit=limit, author_id=author_id)
    total_pages = (total + limit - 1) // limit
    # The top-level author envelope lets the archive page title itself by pen
    # name even when the writer has no published posts yet (DEC-359/TASK-405).
    response = schemas.AuthorPostsResponse.model_validate(
        {
            "items": posts,
            "pagination": {
                "total": total,
                "page": page,
                "limit": limit,
                "total_pages": total_pages,
            },
            "author": author,
        }
    )
    return conditional_json(response.model_dump(mode="json"), request)
